analyses/cms-open-data-ttbar/validate_histograms.py

```diff
@@ -0,0 +1,74 @@
+# Compare the content of histograms produced by ttbar_analysis_pipeline with a reference file.
+# A reference file for N_FILES_MAX_PER_SAMPLE=1 is available in directory `reference/`.
+
+import argparse
+from collections import defaultdict
+import json
+import numpy as np
+import sys
+import uproot
+
+def parse_args() -> argparse.Namespace:
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--histos", help="ROOT file containing the output histograms. Defaults to './histograms.root'.", default="histograms.root")
+    group = parser.add_mutually_exclusive_group(required=True)
+    group.add_argument("--reference", help="JSON reference against which histogram contents should be compared")
+    group.add_argument("--dump-json", help="Print JSON representation of histogram contents to screen", action='store_true')
+    return parser.parse_args()
+
+# convert uproot file containing only TH1Ds to a corresponding JSON-compatible dict with structure:
+# { "histo1": { "edges": [...], "contents": [...] }, "histo2": { ... }, ... }
+# Only the highest namecycle for every histogram is considered, and cycles are stripped from the histogram names.
+def as_dict(f: uproot.ReadOnlyDirectory) -> dict[str, dict]:
+    histos = defaultdict(dict)
+    cycles: dict(str, int) = {}
+    for k, v in f.items():
+        assert isinstance(v, uproot.behaviors.TH1.Histogram)
+        # this assumes that the rightmost ";" (if any) comes before a namecycle
+        name, *cycle = k.rsplit(";", 1)
+        cycle = int(cycle[0]) if len(cycle) > 0 else -1
+        if name in histos and cycles["cycle"] > cycle:
+            continue # found a lower cycle for a histogram we already have
```

Comment:
I think `uproot` would by default just use the highest/latest cycle, so we might not need that, but not 100% sure (cc @jpivarski). On the other hand, I cannot think spontaneously of any harm caused by it either.

Reply:
At least when looping over keys, uproot exposes all namecycles:

```python
In [2]: f = uproot.open("f.root")

In [3]: f.keys()
Out[3]: ['h;2', 'h;1']

In [4]: dict(f.items())
Out[4]:
{'h;2': <TH1D (version 3) at 0x7f5730bb7950>,
 'h;1': <TH1D (version 3) at 0x7f56df9d3fd0>}
```

The removal of the `;N` part of the histogram name is probably desirable anyways.

Reply:
Right, I'm not aware of a way to get this list without the cycles included in the names. For simplicity maybe easiest to just strip the cycle name and then proceed by lookup without cycle names?

Reply:
that requires one loop to get all names (stripping the cycles as they come in) and then another loop to redo the lookup into the file.

nothing here is performance critical so I can do that if you want?

Reply:
I'm not worried about performance, it just seemed to me that we can simplify this:
```python
def as_dict(f: uproot.ReadOnlyDirectory) -> dict[str, dict]:
    histos = defaultdict(dict)
    for k, v in f.items():
        assert isinstance(v, uproot.behaviors.TH1.Histogram)
        # this assumes that the rightmost ";" (if any) comes before a namecycle
        name = k.rsplit(";", 1)[0]
        histos[name]["edges"] = v.axis().edges().tolist()
        histos[name]["contents"] = v.counts(flow=True).tolist()
    return histos
```
Given that `uproot` will automatically pick the latest cycle, I think this is equivalent and allows to drop the manual handling of it?

Reply:
ah, no that's not equivalent, `for k, v in f.items()` loops over _all_ cycles and you'll end up with the contents of _one of those_ in `histos[name]` (not sure which, I guess either the highest or the lowest cycle, but I'm not sure it's guaranteed to be stable) at the end of the loop.

Only accessing a key via `f["histo_name"]` returns the highest cycle.

What I meant above is that accessing the uproot keys twice instead of once we can write this:

```python
def as_dict(f: uproot.ReadOnlyDirectory) -> dict[str, dict]:
    histos = defaultdict(dict)
    # this assumes that the rightmost ";" (if any) comes before a namecycle
    names = set(k.rsplit(";", 1)[0] for k in f)
    for name in names:
        h = f[name]
        assert isinstance(h, uproot.behaviors.TH1.Histogram)
        histos[name]["edges"] = h.axis().edges().tolist()
        histos[name]["contents"] = h.counts(flow=True).tolist()
    return histos
```

and now that I see it written down I do like it better than the current code :smile: 

What do you think?

Reply:
You're right of course, and yes I like that latest version too. Let's use that and get this merged. :)

Reply:
done